h2o-py/h2o/frame.py

```diff
@@ -103,12 +103,23 @@ def __init__(self, python_obj=None, destination_frame=None, header=0, separator=
         assert_is_type(na_strings, None, [str], [[str]], {str: [str]})
         check_frame_id(destination_frame)
 
-        self._ex = ExprNode()
-        self._ex._children = None
         self._is_frame = True  # Indicate that this is an actual frame, allowing typechecks to be made
-        if python_obj is not None:
-            self._upload_python_object(python_obj, destination_frame, header, separator,
-                                       column_names, column_types, na_strings, skipped_columns, force_col_types)
+        if isinstance(python_obj, H2OFrame):
+            sc = h2o.h2o.shallow_copy(python_obj, destination_frame)
+            if skipped_columns:
```

Comment:
@sebhrusen :  I am confused here.  The user want to convert a panda frame to H2O frame.  If the frame is already H2O frame, should it be easier just to return it?  I was thinking this is just a simple check for as_data_frame method....

Reply:
if the user is using a constructor like `H2OFrame`, I don't think it's a good idea to return the exact same frame. That's why I return a shallow copy instead: it's cheap (no data copy) but it's not the same object.

Reply:
You worried about user changing the original frame.

Reply:
> You worried about user changing the original frame.

it's one concern, but not the major one.
I'm more concerned about user expectations: `H2OFrame(…)` is a constructor, not a factory, and therefore I think it should always return a fresh instance.
It could be implemented as a factory but it's not:
- if you call `H2OFrame([1, 2, 3])` twice, you get 2 different objects.
- if you call `H2OFrame(h2o_fr, skipped_columns=['foo', 'bar'])`, you expect a filtered version of the given frame and it will be a new object.
- if you call `H2OFrame(h2o_fr, skipped_columns=['do_not_exist'])`, you expect the same as above, but as the columns don't exist, you will get a new copy of the given frame.
- therefore, if you just call `H2OFrame(h2o_fr)` it's more natural to still return a new instance, and therefore a copy (shallow because there's no reason to deep clone here). If we were returning the given frame directly here, it would be an unnecessary special case that may confuse the user.

Reply:
@wendycwong 
> I am confused here. The user want to convert a panda frame to H2O frame

well, that's what he wants to do when passing a panda frame, but who knows what he wants to do when passing a H2OFrame? It's an edge case.

Maybe to convince you, here is how pandas itself behaves:
```python
In [24]: import pandas as pd

In [25]: df = pd.DataFrame([1, 2, 3])

In [26]: id(df)
Out[26]: 5090919696

In [27]: df2 = pd.DataFrame([1, 2, 3])

In [28]: id(df2)
Out[28]: 5061990032

In [29]: df3 = pd.DataFrame(df)

In [30]: id(df3)
Out[30]: 5043948112
```

you see: `df3 != df`